pkg/plugin/plugin.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+)
+
+// RevaPlugin represents the runtime plugin
+type RevaPlugin struct {
+	Plugin interface{}
+	Client *plugin.Client
+}
+
+// Kill kills the plugin process
+func (plug *RevaPlugin) Kill() {
+	plug.Client.Kill()
+}
+
+var handshake = plugin.HandshakeConfig{
+	ProtocolVersion:  1,
+	MagicCookieKey:   "BASIC_PLUGIN",
+	MagicCookieValue: "hello",
+}
+
+func compile(pluginType string, path string) (string, error) {
+	dirname := "/var/tmp/reva"
+	binaryPath := filepath.Join(dirname, "bin", pluginType, filepath.Base(path))
+	command := fmt.Sprintf("go build -o %s %s", binaryPath, path)
+	cmd := exec.Command("bash", "-c", command)
+	err := cmd.Run()
+	if err != nil {
+		return "", err
+	}
+	return binaryPath, nil
+}
+
+// checkDir checks and compiles plugin if the configuration points to a directory.
+func checkDirAndCompile(pluginType, driver string) (string, error) {
+	bin := driver
+	file, err := os.Stat(driver)
+	if err != nil {
+		return "", err
+	}
+	// compile if we point to a package
+	if file.IsDir() {
+		bin, err = compile(pluginType, driver)
+		if err != nil {
+			return "", err
+		}
+	}
+	return bin, nil
+}
+
+// Load loads the plugin using the hashicorp go-plugin system
+func Load(driver string, pluginType string) (*RevaPlugin, error) {
+	bin, err := checkDirAndCompile(pluginType, driver)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:   "plugin",
+		Output: os.Stdout,
```

Comment:
I don't see the plugin logs anywhere. What do we need to enable that?

Reply:
Ideally you should see the log because I've set log level to `hclog.Trace`. I am getting the following output:
![Screenshot from 2021-07-26 21-18-37](https://user-images.githubusercontent.com/47107987/127019354-c59622b4-e30d-415d-84d1-d7bed2d87bb7.png)

Reply:
I meant how do we log something from the plugin? We probably need to pass the client's context here as our logger is associated with it.

https://github.com/cs3org/reva/blob/968fd84137744f7dd3a52685bc8d0c1785a2e52c/pkg/user/rpc_user.go#L180-L184

Reply:
Done! :+1: